### PR TITLE
test: consistent appHash between commits

### LINKED ISSUE #2414: Non-determinism unit test
## Description

Write a non-determinism unit test for the state machine. The test will involve creating a set of transactions, executing them, and comparing the resulting apphash with an expected value.

## Acceptance Criteria

- [ ] Submit and execute txs using txsim
- [ ] Should cover v1 and v2
- [ ] Assert expected state and data roots

### PR DESCRIPTION
**Pull request by ninabarbakadze**
## Overview

Fixes #2414 

### This test is split up in a few parts:

- Execute transactions deterministically on v1 and save the AppHash. https://github.com/celestiaorg/celestia-app/pull/3522 ([includes genesis folder backport from utils](https://github.com/celestiaorg/celestia-app/pull/3520))
- Set up the same testApp environment on main, execute transactions and assert that the AppHash matches the one from v1

### Verifying the AppHash: 

If you want to make sure that the expected AppHash is correct, you can run the [non determinism test on v1.x](https://github.com/celestiaorg/celestia-app/pull/3522) that generates it.

### Follow-ups:
#3509 
https://github.com/celestiaorg/celestia-app/issues/3540

### CODE REVIEW

**Review by rach-id — DISMISSED** [review]
LGTM, left some optional nits

File: app/test/non_determinism_test.go
Comment (ninabarbakadze) [reply app/test/non_determinism_test.go:30]: good point i'll link the executable from v1.x in the comments

File: app/test/non_determinism_test.go
Comment (rootulp) [inline app/test/non_determinism_test.go:92]:
```suggestion
	assert.Equal(t, expectedAppHash, appHash)
```

**Review by rootulp — COMMENTED** [review]
[future issue / PR] This is a good starting point. I think we could improve this test by executing transaction types that include more unique messages because AFAIU this only tests MsgSend, MsgPayForBlob.

File: app/test/non_determinism_test.go
Comment (ninabarbakadze) [reply app/test/non_determinism_test.go:118]: I was following the convention from our code https://github.com/celestiaorg/celestia-app/blob/fd4b75bb6145da14bdb62476ff065ade9530d0ea/test/util/testnode/utils.go#L45

Comment (ninabarbakadze) [reply app/test/non_determinism_test.go:118]: this function is pretty much identical

File: app/test/non_determinism_test.go
Comment (ninabarbakadze) [reply app/test/non_determinism_test.go:54]: not if you have any randomness like a random namespace every time a test is run. I just wanted to imply that in the comment for anyone debugging this test in the future.

File: app/test/non_determinism_test.go
Comment (ninabarbakadze) [reply app/test/non_determinism_test.go:39]: I try to use the same conventions as our other testing infra https://github.com/celestiaorg/celestia-app/blob/fd4b75bb6145da14bdb62476ff065ade9530d0ea/test/util/testfactory/common.go#L69

File: app/test/non_determinism_test.go
Comment (ninabarbakadze) [reply app/test/non_determinism_test.go:67]: how do you mean different transactions? I'm just using different addresses for blob txs. I just want to start the slice at index 6

File: app/test/non_determinism_test.go
Comment (rootulp) [reply app/test/non_determinism_test.go:118]: kk can resolve. I have no idea what `kb` means though

File: app/test/non_determinism_test.go
Comment (rootulp) [reply app/test/non_determinism_test.go:92]: Thanks, updating the comment so that they both say assert or require would be nice. Since there are no more assertions after this line an assert will behave identically to a require.

File: app/test/non_determinism_test.go
Comment (rootulp) [reply app/test/non_determinism_test.go:39]: Fair enough, can preserve that for now. In a separate PR if we really care, we can rename recs => records

**Comment by cmwaters** [comment]
> I think we could improve this test by executing transaction types that include more unique messages because AFAIU this only tests MsgSend, MsgPayForBlob.

I agree but I think first merge this with the basic structure and then we can further elaborate

File: app/test/non_determinism_test.go
Comment (cmwaters) [inline app/test/non_determinism_test.go:25]: This shouldn't necessary be between v1 and main. It should be comparing the app hash generated from this commit compared to the app hash generated from the previous commit

**Review by cmwaters — COMMENTED** [review]
I would have constructed this test a little differently. I would use a single `user.Signer` that has the keyring with all accounts. Then I would have something like.

```go
txs := []struct {
    sdkMsgs []sdk.Msg
    author string
    blobs []blob.Blob
    txOptions []user.TxOption
}{
   // add a mixture of blob txs and non blob txs with various signers
}
// you may want to break this out over multiple blocks
txBytes := make([]byte, len(txs))
for idx, tx := range txs {
    if len(tx.sdkMsgs) > 0 {
        txBytes[idx] = signer.CreateTx(tx.sdkMsgs, tx.txOptions)
    } else {
       txBytes[idx] = signer.CreatePayForBlob(tx.author, tx.blobs, tx.txOptions)
   }
}
// run prepare proposal and process proposal (make sure data hashes are deterministic)

// run `BeginBlock`, `DeliverTx`,  and `EndBlock` with the txBytes

// run `Commit` and get the app hash. Compare with the static one

```

File: app/test/non_determinism_test.go
Comment (evan-forbes) [reply app/test/non_determinism_test.go:25]: agree that I think we can rename this function since the commitment should be the same across all branches if I'm understanding this comment correctly.

**Review by evan-forbes — DISMISSED** [review]

File: app/test/non_determinism_test.go
Comment (rootulp) [reply app/test/non_determinism_test.go:25]: [optional] slight tangent: the test could be renamed something like `TestConsistentAppHash` because it is asserting some expected app hash matches some calculated app hash. I think the goal is to identify cases of non-determinism but just because this test passes, it doesn't mean we prevented all potential cases of non-determinism.

Comment (ninabarbakadze) [reply app/test/non_determinism_test.go:25]: i agree with all of your point, i only wanted to underline that the hardcoded apphash was generated on v1.x because technically we could do it between any branches but practically the logic for generating apphash lives on v1.x.

File: app/test/non_determinism_test.go
Comment (ninabarbakadze) [reply app/test/non_determinism_test.go:30]: I'll link the file once this pr is merged as a follow-up https://github.com/celestiaorg/celestia-app/pull/3522

File: app/test/consistent_apphash_test.go
Comment (rootulp) [inline app/test/consistent_apphash_test.go:137]: 1. comment name should match function name
2. If this isn't used outside this test file, proposal to un-export it.

```suggestion
// namespace returns a deterministic namespace
func namespace() appns.Namespace {
```

File: app/test/consistent_apphash_test.go
Comment (rootulp) [inline app/test/consistent_apphash_test.go:145]: [optional] un-export

```suggestion
// deterministicKeyRing returns a deterministic keyring and a list of deterministic public keys
func deterministicKeyRing(cdc codec.Codec) (keyring.Keyring, []types.PubKey) {
```

File: test/util/test_app.go
Comment (rootulp) [inline test/util/test_app.go:169]: [nit] since this function doesn't take a validator set as a parameter, I think the current godoc is misleading.

```suggestion
// NewTestAppWithGenesisSet initializes a new app with genesis accounts and returns the testApp, validator set and keyring.
```

File: test/util/testnode/utils.go
Comment (rootulp) [inline test/util/testnode/utils.go:68]: [informational] note this modification is technically Go API breaking. We technically don't need to preserve Go API compatability because this is targeting `main` so FYI but no change needed

**Review by rootulp — APPROVED** [review]

File: test/util/test_app.go
Comment (ninabarbakadze) [reply test/util/test_app.go:169]: i was not only referring to the arguments it takes but also the overall functionality of the method and what it initialises. happy to change it if you think it's still misleading.

File: app/test/consistent_apphash_test.go
Comment (ninabarbakadze) [reply app/test/consistent_apphash_test.go:137]: i can't just have namespace because it conflicts with the package name but i renamed it to `fixedNamespace`

**Review by evan-forbes — APPROVED** [review]

File: test/util/test_app.go
Comment (cmwaters) [inline test/util/test_app.go:98]: why not?
```suggestion
func ApplyGenesisState(testApp *app.App, gen *genesis.Genesis) error {
```
The way you currently have it is not applying genesis but creating a genesis.

File: test/util/test_app.go
Comment (cmwaters) [inline test/util/test_app.go:213]: If you look at `test/util/common.go`, they already have a bunch of keys set up that you can use

File: test/util/testnode/utils.go
Comment (cmwaters) [reply test/util/testnode/utils.go:68]: I would prefer if we add this as a completely new function called `FundAddresses` since it's quite a different function

File: test/util/test_app.go
Comment (ninabarbakadze) [reply test/util/test_app.go:98]: preferably i'd have a function that lets you customise genesis without having to do the setup yourself. i'm happy to change the function signature to correctly reflect the behaviour. we also have [a ticket refactoring test_app ](https://github.com/celestiaorg/celestia-app/issues/2535) so we'll spend more time properly designing this process.

File: test/util/test_app.go
Comment (ninabarbakadze) [reply test/util/test_app.go:213]: I don't think those keys are deterministic

File: test/util/testnode/utils.go
Comment (ninabarbakadze) [reply test/util/testnode/utils.go:68]: I'll revert this change completely. I thought that decoupling those functions made sense but we don't really use them elsewhere.

File: test/util/test_app.go
Comment (ninabarbakadze) [reply test/util/test_app.go:98]: technically this method is applying genesis on already initialised testapp

File: test/util/test_app.go
Comment (rootulp) [reply test/util/test_app.go:169]: I think it's misleading so I prefer the comment I proposed but not blocking on it.

**Review by rootulp — APPROVED** [review]

**Review by evan-forbes — APPROVED** [review]

**Pull request merged**
